### ISSUE TITLE
[resize-observer] resize-observer/notify.html should throw a "undelivered notifications" error?

### ISSUE THREAD

**Issue by smfr** [web-platform-tests/wpt#30708]
I'm fairly sure that, per https://drafts.csswg.org/resize-observer/, resize-observer/notify.html should throw a "undelivered notifications" in `test8()`.

The test does this:
```
    {
      setup: observer => {
        observer.observe(t1);
      },
      notify: (entries, observer) => {
        observer.observe(t2);
        t2.style.width = "888px";
        return true;  // Delay next step
      }
    },
```

`notify` is should be called three times; once for t1, and twice for t2: one time when t2 is first observed, and then (per spec), because `observe(t2)` calls `unobserve(t2)` (https://drafts.csswg.org/resize-observer/#dom-resizeobserver-observe) and then re-observes, resulting in the third observation callback. However, t2 is at level 4 both times, so the second callback is skipped and should result in a "undelivered notifications" error.

Current implementations seem to avoid this, possibly by just early-returning from `observe()` when the target is the existing target and the observed box is the same.

cc @atotic

**Comment by smfr** [comment]
Several tests in observe.html also have the same issue, like `test5()`. Blink's implementation seems to be subtly different from the spec in a way that avoids this.

**Comment by atotic** [comment]
> `notify` is should be called three times; once for t1, and twice for t2: one time when t2 is first observed, and then (per spec), because `observe(t2)` calls `unobserve(t2)` (https://drafts.csswg.org/resize-observer/#dom-resizeobserver-observe) and then re-observes, resulting in the third observation callback. However, t2 is at level 4 both times, so the second callback is skipped and should result in a "undelivered notifications" error.

This is my understanding of what should happen
```
observer.observe(t1);
// after style/layout we start the RO notification loop. t1's size has changed.
notify([t1]) { // gets called with t1 notification
  // observe spec: If target is in [[observationTargets]] slot, call unobserve(target).
  observer.observe(t2);   // unobserve is not called, because t2 is not inside observationTargets
  t2.style.width = "888px";
}
// RO notification runs again, with increased min-depth 
// this time, only t2's size has changed. t1's size remains the same
notify([t2]) {
  // assert only t2's size has changed...
}
```
I do not see "two calls per t2" scenario. 
> one time when t2 is first observed (yes!)
>   then (per spec), because `observe(t2)` calls `unobserve(t2)

unobserve gets called only if t2 is already being observed.

Am I getting something wrong here?

**Comment by smfr** [comment]
You're right, I was confused about `t1` vs `t2` (although I think that might still be an issue).

But let's say `t1` and `t2` are the at same depth (which I believe they are). We're in the notify for `t1` at depth N and observe and trigger a notification on `t2`. Won't the `t2` notification fail the "If targetDepth is greater than depth" check and therefore trigger a skipped notification?

`test5()` in observer.html does something similar: 
```
      setup: observer => {
        img.style.width = "15.5px";
      },
```
which I think should trigger another notification at the same depth, and therefore cause a skipped notification.

**Comment by smfr** [comment]
I think this would trigger the same issue:
```
let element = ...
let observer = new ResizeObserver(entries => {
  observer.observe(element);
}).observe(element);
```

because `observe(element)` the second time triggers `unobserve` followed by `observe`, which will add a new entry which will then try to fire at the same depth and should then caused a skipped notification.

**Comment by atotic** [comment]
Your last example (simple observe inside notification) is correct, and per spec.

Chrome has an optimization: if you call observer on an already observed element with identical options, that call is ignored.
It seems that FF has the same optimization (and it seems like a good developer ergonomics).

Both Chrome and FF will fire an error if 2nd observe() is called with different box options.

Maybe we should change the spec to:

let preexistingTarget be a slot in [[observationTargets]] where preexistingTarget.target equals target.
If preexistingTarget exists and its observedBox.box equals options.box return
if preexistingTarget exists call unobserve(target)

Also, while debugging this I've noticed an existing Chrome bug where RO errors are not printed to the console crbug.com/1040031 (just in case you are investigating too).

**Comment by smfr** [comment]
> Both Chrome and FF will fire an error if 2nd observe() is called with different box options.

Would be good to have a WPT for that (I didn't notice one).

It will be interesting to see if Chrome is actually issuing the same RO errors (and whether doing so change WPT behavior).

**Comment by atotic** [comment]
I just typed in this long response about how observe.html:test5 is correct, and I've now figured out what the bug is.

The test assumes that setup1 and setup2 are called inside two different runs of the event loop.
But this is not what happens in this test. That might be why 
`setup({allow_uncaught_exception: true}); ` is at the top of the file.

To make 2nd setup happen after 1st event loop, 1st notify should return true.
If notify returns true, ResizeTestHelper calls next setup after existing event loop is complete.

The test would still pass, but it would have triggered an error handler.
